### PR TITLE
Add the ability to add and remove bindings in subtrees rooted at a specified node

### PR DESCRIPTION
**Pull request by kmh287**
- Adds the ability to add and remove bindings in subtrees rooted at a specific node
- Adds tests for the new behavior

/to @choumx

### CODE REVIEW

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -142,16 +147,27 @@ export class Bind {
...
+      const {boundElements, bindings, expressionToElements} = results;
 
+      this.boundElements_ = this.boundElements_.concat(boundElements);
       Object.assign(this.expressionToElements_, expressionToElements);
+      this.bindings_ = this.bindings_.concat(bindings);
 
-      this.evaluator_ = new BindEvaluator();
-      const parseErrors = this.evaluator_.setBindings(bindings);
+      this.evaluator_ = this.evaluator_ || new BindEvaluator();
+      const parseErrors = this.evaluator_.setBindings(this.bindings_);
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: This will cause the evaluator to re-parse every expression, and expression parsing currently is the slowest operation in amp-bind. Instead, we should support incrementally add of expressions.

You can probably avoid the `this.bindings_` variable, which saves work when removing bindings from a subtree.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -170,12 +186,76 @@ export class Bind {
...
+        for (const expression in expressionToElements) {
+          if (expression in this.expressionToElements_) {
+            const elements = this.expressionToElements_[expression];
+            for (let k = elements.length - 1; k >= 0; k--) {
+              if (elements[k].isEqualNode(boundElementToRemove.element)) {
+                elements.splice(k, 1);
+              }
+            }
+            if (elements.length == 0) {
+              delete this.expressionToElements_[expression];
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: We should also remove obsolete expressions from the evaluator.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 3]
```diff
@@ -170,12 +186,76 @@ export class Bind {
...
+  /**
+   * Scans the substree rooted at `rootElement` and removes bindings for nodes
+   * that contain bindable elements. This function is not idempotent. To add
+   * bindings for a subtree, see #addBindingsForSubtree. Returns a promise
+   * that resolves after bindings have been removed.
+   * @param {!Element} rootElement
+   * @return {Promise}
+   */
+  removeBindingsForSubtree(rootElement) {
+    this.scanPromise_ = this.scanSubtree_(rootElement).then(results => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: I don't think we need to scan the subtree here since we should already have the bindings/expression strings that we're about to remove.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 4]
```diff
@@ -170,12 +186,76 @@ export class Bind {
...
+      const {boundElements, bindings, expressionToElements} = results;
+
+      // TODO(kmh287): Discuss strategies for speedup
+      for (let i = 0; i < boundElements.length; i++) {
+        const boundElementToRemove = boundElements[i];
+        for (let j = this.boundElements_.length - 1; j >= 0; j--) {
+          const currentElement = this.boundElements_[j];
+          if (currentElement
+                .element
+                .isEqualNode(boundElementToRemove.element)) {
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: We want to match the exact node, not just equivalent nodes. Otherwise, we may remove bindings from elements outside of the subtree.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 5]
```diff
@@ -170,12 +186,76 @@ export class Bind {
...
+   * bindings for a subtree, see #addBindingsForSubtree. Returns a promise
+   * that resolves after bindings have been removed.
+   * @param {!Element} rootElement
+   * @return {Promise}
+   */
+  removeBindingsForSubtree(rootElement) {
+    this.scanPromise_ = this.scanSubtree_(rootElement).then(results => {
+      const {boundElements, bindings, expressionToElements} = results;
+
+      // TODO(kmh287): Discuss strategies for speedup
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: You can probably just walk the ancestor list of each bound element and check whether `rootElement` is one of them.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 6]
```diff
@@ -142,16 +147,27 @@ export class Bind {
...
-      this.boundElements_ = boundElements;
+  /**
+   * Scans the substree rooted at `rootElement` and adds bindings for nodes
+   * that contain bindable elements. This function is not idempotent. To remove
+   * bindings for a subtree, see #removeBindingsForSubtree. Returns a promise
+   * that resolves after bindings have been added.
+   * @param {!Element} rootElement
+   * @return {Promise}
+   */
+  addBindingsForSubtree(rootElement) {
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Nit: s/Subtree/Node and s/rootElement/node

Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Mark all private methods with `@private` attribute and trailing underscore. This allows Closure compiler to rename the method during minimization and other optimizations like DCE.

File: extensions/amp-bind/0.1/test/test-bind-impl.js [hunk 7]
```diff
@@ -76,39 +79,33 @@ describes.realWin('amp-bind', {
 
   /**
    * Calls `callback` when Bind's DOM scan and optional verify completes.
-   * @param {!Function} callback
    * @return {!Promise}
    */
-  function onBindReady(callback) {
+  function onBindReady() {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: 👍

File: extensions/amp-bind/0.1/test/test-bind-impl.js [hunk 8]
```diff
@@ -119,112 +116,142 @@ describes.realWin('amp-bind', {
     }).to.throw('Experiment "amp-bind" is disabled.');
   });
 
-  it('should scan for bindings when ampdoc is ready', () => {
+  it('should scan for bindings when ampdoc is ready', done => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Remove unused `done` here and below.

File: extensions/amp-bind/0.1/test/test-bind-impl.js [hunk 9]
```diff
@@ -56,9 +56,12 @@ describes.realWin('amp-bind', {
    * @return {!Element}
    */
   function createElementWithBinding(binding) {
+    const fakeDiv = env.win.document.createElement('div');
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Why this change? Also, is this div actually "fake"?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 4 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Will exactly the same node always be represented by the same object? i.e. will === work?

File: extensions/amp-bind/0.1/test/test-bind-impl.js (hunk 9 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: The old version only ever created one element with a binding. I needed to change it to make one per call.

File: extensions/amp-bind/0.1/test/test-bind-impl.js (hunk 8 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Now that onBindReady() returns a promise, it looks like we need this to signal the tests are done after the promises resolve.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 6 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done

File: extensions/amp-bind/0.1/bind-impl.js (hunk 2 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 4 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: https://developer.mozilla.org/en-US/docs/Web/API/Node/isSameNode

File: extensions/amp-bind/0.1/test/test-bind-impl.js (hunk 8 above)
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Shouldn't be necessary. Grab me if it's not working.

Comment (jridgewell) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Mocha understand promise return values, so there's no need for done.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 10]
```diff
@@ -170,12 +183,74 @@ export class Bind {
...
+    });
+  }
+
+  /**
+   * Checks if `potentialParent` is an ancestor node of `node`
+   * @param {!Element} node
+   * @param {!Element} potentialParent
+   * @return {boolean}
+   */
+  isAncestorOf_(node, potentialParent) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Nodes have a [_super_ efficient](https://jsperf.com/node-contains-vs-while-loop/6) `#contains` to do this:

```js
isAncestorOf_(node, potentialParent) {
  return potentialParent.contains(node);
}

File: extensions/amp-bind/0.1/bind-impl.js [hunk 11]
```diff
@@ -170,12 +183,74 @@ export class Bind {
...
+   *
+   * @param {!Element} node
+   * @return {Promise}
+   *
+   * @private
+   */
+  removeBindingsForNode_(node) {
+    return new Promise(resolve => {
+      // Eliminate bound elements that have node as an ancestor
+      for (let i = this.boundElements_.length - 1; i >= 0; i--) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: We have a [`filterSplice`](https://github.com/ampproject/amphtml/blob/master/src/utils/array.js#L26) helper for this.

File: extensions/amp-bind/0.1/test/test-bind-impl.js (hunk 8 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 11 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 10 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Awesome. Done.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 12]
```diff
@@ -80,6 +81,24 @@ export class BindEvaluator {
...
+   * Removes all parsed bindings for the provided expressions.
+   * @param {!Array<string>} expressionStrings
+   */
+  removeBindingsForExpressions(expressionStrings) {
+    const expressionsToRemove = map();  // Used as set
+    for (let i = 0; i < expressionStrings.length; i++) {
+      expressionsToRemove[expressionStrings[i]] = undefined;
+    }
+
+    for (let j = this.parsedBindings_.length - 1; j >= 0; j--) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Ditto `filterSplice`

File: extensions/amp-bind/0.1/bind-impl.js [hunk 13]
```diff
@@ -170,12 +184,71 @@ export class Bind {
...
+        return node.contains(boundElement.element);
+      });
+
+      // Eliminate elements from the expression to elements map that
+      // have node as an ancestor. Delete expressions that are no longer
+      // bound to elements.
+      const deletedExpressions = [];
+      for (const expression in this.expressionToElements_) {
+        const elements = this.expressionToElements_[expression];
+        for (let j = elements.length - 1; j >= 0; j--) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Ditto.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 10 above)
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: This is no longer called, right? Please delete.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    const bindingDef2 = {
+      tagName: 'SPAN',
+      property: 'text',
+      expressionString: 'oneplusone + 3',
+    };
+    expect(evaluator.parsedBindings_.length).to.equal(0);
+    evaluator.addBindings([bindingDef]);
+    expect(evaluator.parsedBindings_.length).to.equal(1);
+    evaluator.addBindings([bindingDef2]);
+    expect(evaluator.parsedBindings_.length).to.equal(2);
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Ideally, unit tests should only test public interfaces. For this case I think it's ok, though exposing `parsedBindings_` via a new method `BindEvaluator#parsedBindingsForTesting` is more refactor-friendly than referencing the ivar directly.

I noticed that I didn't follow my own advice in `test-bind-impl.js`. 😞

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    evaluator.removeBindingsForExpressions([bindingDef.expressionString]);
+    expect(evaluator.parsedBindings_.length).to.equal(1);
+    evaluator.removeBindingsForExpressions([bindingDef2.expressionString]);
+    expect(evaluator.parsedBindings_.length).to.equal(0);
+  });
+
+  it('should evaluate expressions given a scope with needed bindings', done => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: This test's correctness heavily depends on `BindExpression`'s correctness, which makes it more of an integration test than a unit test. `test-bind-impl.js` already contains a few tests like this -- we should probably extract to a discrete `test-bind-integration.js` test file.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+      const evaluated = results['results'];
+      const errors = results['errors'];
+      expect(errors[bindingDef.expressionString]).to.be.undefined;
+      expect(evaluated[bindingDef.expressionString]).to.not.be.undefined;
+      expect(evaluated[bindingDef.expressionString] = '4');
+      done();
+    });
+  });
+
+  it('should treat out-of-scope vars as null', done => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Ditto.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 17]
```diff
@@ -170,12 +184,52 @@ export class Bind {
...
+        filterSplice(elements, element => {
+          return !node.contains(element);
+        });
+        if (elements.length == 0) {
+          deletedExpressions.push(expression);
+          delete this.expressionToElements_[expression];
+        }
+      }
+
+      // Remove the bindings from the evaluator
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Ditto.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 18]
```diff
@@ -80,6 +83,22 @@ export class BindEvaluator {
   }
 
   /**
+   * Removes all parsed bindings for the provided expressions.
+   * @param {!Array<string>} expressionStrings
+   */
+  removeBindingsForExpressions(expressionStrings) {
```
Comment (choumx) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Nit: `removeBindingsWithExpressionStrings`.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BindEvaluator, BindingDef} from '../bind-evaluator';
+
+describe('BindEvaluator', () => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Thanks for adding tests to this class!

File: extensions/amp-bind/0.1/test/test-bind-impl.js [hunk 20]
```diff
@@ -122,15 +119,33 @@ describes.realWin('amp-bind', {
...
 
+  it('should have same state after removing + re-adding a subtree', () => {
+    for (let i = 0; i < 5; i++) {
+      createElementWithBinding('[onePlusOne]="1+1"');
+    }
+    expect(bind.boundElements_.length).to.equal(0);
+    return onBindReady().then(() => {
+      expect(bind.boundElements_.length).to.equal(5);
+      return bind
+        .removeBindingsForNode_(env.win.document.getElementById('parent'));
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Nit: Move `env.win.document` to a local var and this can fit on one line.

**Comment by kmh287** [comment]
@choumx I've merged in the web worker. PTAL

File: extensions/amp-bind/0.1/test/test-bind-impl.js [hunk 21]
```diff
@@ -122,15 +119,33 @@ describes.realWin('amp-bind', {
   it('should scan for bindings when ampdoc is ready', () => {
     createElementWithBinding('[onePlusOne]="1+1"');
     expect(bind.boundElements_.length).to.equal(0);
-    return onBindReady(() => {
+    return onBindReady().then(() => {
       expect(bind.boundElements_.length).to.equal(1);
     });
   });
 
+  it('should have same state after removing + re-adding a subtree', () => {
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: Recommend also verifying invocation of corresponding `BindEvaluator` methods.

Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-impl.js]: IMO the method invocations on the evaluator are more of an implementation detail of the Bind object.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js (hunk 16 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    };
+    expect(evaluator.parsedBindings_.length).to.equal(0);
+    evaluator.addBindings([bindingDef]);
+    expect(evaluator.parsedBindings_.length).to.equal(1);
+    evaluator.addBindings([bindingDef2]);
+    expect(evaluator.parsedBindings_.length).to.equal(2);
+  });
+
+  it('should allow callers to remove bindings', () => {
+    const bindingDef = {
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js (hunk 14 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/test/test-bind-evaluator.js [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+describe('BindEvaluator', () => {
+
+  let evaluator;
+
+  beforeEach(() => {
+    evaluator = new BindEvaluator();
+  });
+
+  it('should allow callers to add bindings multiple times', () => {
+    const bindingDef = {
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 24]
```diff
@@ -80,6 +83,22 @@ export class BindEvaluator {
...
+    const expressionsToRemove = map();  // Used as set
+    for (let i = 0; i < expressionStrings.length; i++) {
+      expressionsToRemove[expressionStrings[i]] = undefined;
+    }
+
+    this.parsedBindings_ = filterSplice(this.parsedBindings_, binding => {
+      const expression = binding.expression.expressionString;
+      return !hasOwn(expressionsToRemove, expression);
```
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 25]
```diff
@@ -80,6 +83,22 @@ export class BindEvaluator {
...
+   * @param {!Array<string>} expressionStrings
+   */
+  removeBindingsForExpressions(expressionStrings) {
+    const expressionsToRemove = map();  // Used as set
+    for (let i = 0; i < expressionStrings.length; i++) {
+      expressionsToRemove[expressionStrings[i]] = undefined;
+    }
+
+    this.parsedBindings_ = filterSplice(this.parsedBindings_, binding => {
+      const expression = binding.expression.expressionString;
```
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/bind-evaluator.js (hunk 18 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 17 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 26]
```diff
@@ -170,12 +184,52 @@ export class Bind {
...
+   *
+   * @param {!Element} node
+   * @return {Promise}
+   *
+   * @private
+   */
+  removeBindingsForNode_(node) {
+    return new Promise(resolve => {
+
+      // Eliminate bound elements that have node as an ancestor
```
Comment (kmh287) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 27]
```diff
@@ -79,6 +80,22 @@ export class BindEvaluator {
...
+   */
+  removeBindingsWithExpressionStrings(expressionStrings) {
+    const expressionsToRemove = Object.create(null);
+    for (let i = 0; i < expressionStrings.length; i++) {
+      expressionsToRemove[expressionStrings[i]] = true;
+    }
+
+    this.parsedBindings_ = filterSplice(this.parsedBindings_, binding => {
+      const expressionString = binding.expression.expressionString;
+      return expressionsToRemove[expressionString];
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-evaluator.js]: You want `!` this. `#filterSplice` (and `#filter`) keep the items that return true and remove the items that return false.

Comment (kmh287) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Thanks for catching this.

Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Lol.

File: src/web-worker/web-worker.js [hunk 28]
```diff
@@ -35,19 +35,26 @@ self.addEventListener('message', function(event) {
...
-      returnValue = evaluator_.setBindings.apply(evaluator_, args);
+    case 'bind.addBindings':
+      evaluator_ = evaluator_ || new BindEvaluator();
+      returnValue = evaluator_.addBindings.apply(evaluator_, args);
+      break;
+    case 'bind.removeBindings':
+      if (evaluator_) {
+        returnValue =
+          evaluator_
+          .removeBindingsWithExpressionStrings.apply(evaluator_, args);
```
Comment (choumx) [inline src/web-worker/web-worker.js]: Move `evaluator_.removeBindingsWithExpressionStrings` to a local var to fix this formatting.

**Comment by kmh287** [comment]
Thank you for the thorough review
